src/EndpointManager.sol

```diff
@@ -241,34 +313,62 @@ abstract contract EndpointManager is
-
-            // query own token balance after transfer
-            uint256 balanceAfter = getTokenBalanceOf(_token, address(this));
-
-            // correct amount for potential transfer fees
-            amount = balanceAfter - balanceBefore;
-        } else {
-            // query sender's token balance before transfer
-            uint256 balanceBefore = getTokenBalanceOf(_token, msg.sender);
-
-            // call the token's burn function to burn the sender's token
-            ERC20Burnable(_token).burnFrom(msg.sender, amount);
-
-            // query sender's token balance after transfer
-            uint256 balanceAfter = getTokenBalanceOf(_token, msg.sender);
+        if (shouldCheckRateLimit) {
+            // Lock/burn tokens before checking rate limits
+            if (_mode == Mode.LOCKING) {
+                // use transferFrom to pull tokens from the user and lock them
+                // query own token balance before transfer
+                uint256 balanceBefore = getTokenBalanceOf(_token, address(this));
+
+                // transfer tokens
+                IERC20(_token).safeTransferFrom(msg.sender, address(this), amount);
+
+                // query own token balance after transfer
+                uint256 balanceAfter = getTokenBalanceOf(_token, address(this));
+
+                // correct amount for potential transfer fees
+                amount = balanceAfter - balanceBefore;
+            } else {
+                // query sender's token balance before transfer
+                uint256 balanceBefore = getTokenBalanceOf(_token, msg.sender);
+
+                // call the token's burn function to burn the sender's token
+                ERC20Burnable(_token).burnFrom(msg.sender, amount);
+
+                // query sender's token balance after transfer
+                uint256 balanceAfter = getTokenBalanceOf(_token, msg.sender);
+
+                // correct amount for potential burn fees
+                amount = balanceAfter - balanceBefore;
+            }
+
+            // now check rate limits
+            bool didConsumeAmount = _consumeOutboundAmount(amount);
+            if (shouldQueue && !didConsumeAmount) {
+                // queue up and return
+                uint64 queueSequence = useOutboundQueueSequence();
+
+                _getOutboundQueueStorage()[queueSequence] = OutboundQueuedTransfer({
+                    amount: amount,
+                    recipientChain: recipientChain,
+                    recipient: recipient,
+                    txTimestamp: block.timestamp
+                });
+
+                // refund the price quote back to sender
+                payable(msg.sender).transfer(msg.value);
+
+                // return the sequence in the queue
+                return queueSequence;
+            } else if (!didConsumeAmount) {
+                revert NotEnoughOutboundCapacity(getCurrentOutboundCapacity(), amount);
```

Comment:
Ah, so the way I wrote this was to have `didConsumeAmount == true` when the transaction can go through, and have `didConsumeAmount == false` when the transaction can't go through (hits a rate limit).

Maybe naming is the issue here -- instead of `didConsumeAmount` perhaps we could name something like `isAmountRateLimited`

Reply:
got it - yea i think `isAmountRateLimited` or `isAtCapacity` would be good. Also, an early return may be good if `shouldQueue` is `false` given the queuing control flow shouldn't happen in that case